File: Basics/graph/graph-DFS-BFS.go
package main

import (
	"fmt"
	"sync"
)

type Graph struct {
	adjacencyList map[string][]string
}

func NewGraph() *Graph {
	return &Graph{adjacencyList: make(map[string][]string)}
}

func (g *Graph) AddVertex(vertex string) {
	g.adjacencyList[vertex] = []string{}
}

func (g *Graph) AddEdge(vertex1, vertex2 string) {
	g.adjacencyList[vertex1] = append(g.adjacencyList[vertex1], vertex2)
	g.adjacencyList[vertex2] = append(g.adjacencyList[vertex2], vertex1)
}

func (g *Graph) DFS(startVertex string, visited map[string]bool, wg *sync.WaitGroup) {
	defer wg.Done()
	fmt.Println(startVertex)
	visited[startVertex] = true

	for _, neighbor := range g.adjacencyList[startVertex] {
		if !visited[neighbor] {
			wg.Add(1)
			go g.DFS(neighbor, visited, wg)
		}
	}
}

func (g *Graph) BFS(startVertex string) {
	queue := []string{startVertex}
	visited := make(map[string]bool)

	for len(queue) > 0 {
		currentVertex := queue[0]
		fmt.Println(currentVertex)
		visited[currentVertex] = true
		queue = queue[1:]

		for _, neighbor := range g.adjacencyList[currentVertex] {
			if !visited[neighbor] {
				queue = append(queue, neighbor)
				visited[neighbor] = true
			}
		}
	}
}

func main() {
	myGraph := NewGraph()
	myGraph.AddVertex("A")
	myGraph.AddVertex("B")
	myGraph.AddVertex("C")
	myGraph.AddEdge("A", "B")
	myGraph.AddEdge("B", "C")

	fmt.Println("DFS:")
	var wg sync.WaitGroup
	wg.Add(1)
	go myGraph.DFS("A", make(map[string]bool), &wg)
	wg.Wait()

	fmt.Println("BFS:")
	myGraph.BFS("A")
}

File: Basics/graph/graph-DFS-BFS.rs
use std::collections::{HashMap, HashSet, VecDeque};

struct Graph {
    adjacency_list: HashMap<String, Vec<String>>,
}

impl Graph {
    fn new() -> Self {
        Graph {
            adjacency_list: HashMap::new(),
        }
    }

    fn add_vertex(&mut self, vertex: &str) {
        self.adjacency_list.insert(vertex.to_string(), Vec::new());
    }

    fn add_edge(&mut self, vertex1: &str, vertex2: &str) {
        self.adjacency_list
            .get_mut(vertex1)
            .unwrap()
            .push(vertex2.to_string());
        self.adjacency_list
            .get_mut(vertex2)
            .unwrap()
            .push(vertex1.to_string());
    }

    fn dfs(&self, start_vertex: &str, visited: &mut HashSet<String>) {
        println!("{}", start_vertex);
        visited.insert(start_vertex.to_string());

        for neighbor in self.adjacency_list.get(start_vertex).unwrap() {
            if !visited.contains(neighbor) {
                self.dfs(neighbor, visited);
            }
        }
    }

    fn bfs(&self, start_vertex: &str) {
        let mut queue = VecDeque::new();
        let mut visited = HashSet::new();

        queue.push_back(start_vertex.to_string());

        while let Some(current_vertex) = queue.pop_front() {
            println!("{}", current_vertex);
            visited.insert(current_vertex.clone());

            for neighbor in self.adjacency_list.get(&current_vertex).unwrap() {
                if !visited.contains(neighbor) && !queue.contains(neighbor) {
                    queue.push_back(neighbor.to_string());
                }
            }
        }
    }
}

fn main() {
    let mut my_graph = Graph::new();
    my_graph.add_vertex("A");
    my_graph.add_vertex("B");
    my_graph.add_vertex("C");
    my_graph.add_edge("A", "B");
    my_graph.add_edge("B", "C");

    println!("DFS:");
    my_graph.dfs("A", &mut HashSet::new());

    println!("BFS:");
    my_graph.bfs("A");
}
